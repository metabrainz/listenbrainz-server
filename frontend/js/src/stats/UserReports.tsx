import { createRoot } from "react-dom/client";
import * as React from "react";

import * as Sentry from "@sentry/react";
import { Integrations } from "@sentry/tracing";
import NiceModal from "@ebay/nice-modal-react";
import { faGlobe, faUser } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import ErrorBoundary from "../utils/ErrorBoundary";
import Pill from "../components/Pill";
import UserListeningActivity from "./UserListeningActivity";
import UserTopEntity from "./UserTopEntity";
import UserDailyActivity from "./UserDailyActivity";
import UserArtistMap from "./UserArtistMap";
import { getPageProps } from "../utils/utils";
import { getAllStatRanges } from "./utils";
import {
  WithAlertNotificationsInjectedProps,
  withAlertNotifications,
} from "../notifications/AlertNotificationsHOC";
import GlobalAppContext from "../utils/GlobalAppContext";

export type UserReportsProps = {
  user?: ListenBrainzUser;
  apiUrl: string;
} & WithAlertNotificationsInjectedProps;

export type UserReportsState = {
  range: UserStatsAPIRange;
};

export default class UserReports extends React.Component<
  UserReportsProps,
  UserReportsState
> {
  static contextType = GlobalAppContext;
  declare context: React.ContextType<typeof GlobalAppContext>;

  constructor(props: UserReportsProps) {
    super(props);

    this.state = {
      range: "" as UserStatsAPIRange,
    };
  }

  componentDidMount() {
    window.addEventListener("popstate", this.syncStateWithURL);

    const range = this.getURLParams();
    window.history.replaceState(
      null,
      "",
      `?range=${range}${window.location.hash}`
    );
    this.syncStateWithURL();
  }

  componentWillUnmount() {
    window.removeEventListener("popstate", this.syncStateWithURL);
  }

  changeRange = (newRange: UserStatsAPIRange): void => {
    this.setURLParams(newRange);
    this.syncStateWithURL();
  };

  syncStateWithURL = async (): Promise<void> => {
    const range = this.getURLParams();
    this.setState({ range });
  };

  getURLParams = (): UserStatsAPIRange => {
    const url = new URL(window.location.href);

    let range: UserStatsAPIRange = "week";
    if (url.searchParams.get("range")) {
      range = url.searchParams.get("range") as UserStatsAPIRange;
    }

    return range;
  };

  setURLParams = (range: UserStatsAPIRange): void => {
    window.history.pushState(null, "", `?range=${range}`);
  };

  render() {
    const { range } = this.state;
    const { apiUrl, user, newAlert } = this.props;
    const { currentUser } = this.context;

    const ranges = getAllStatRanges();
    let userStatsUrl: string | undefined;
    if (user?.name) {
      userStatsUrl = `${window.location.origin}/user/${user.name}/stats/?range=${range}`;
    } else if (currentUser?.name) {
      userStatsUrl = `${window.location.origin}/user/${currentUser.name}/stats/?range=${range}`;
    }

    const globalStatsUrl = `${window.location.origin}/statistics/?range=${range}`;

    return (
      <div>
        <div className="row mt-15 flex flex-wrap">
          <div style={{ flex: "initial" }}>
            {Array.from(ranges, ([stat_type, stat_name]) => {
              return (
                <Pill
                  active={range === stat_type}
                  type="secondary"
                  onClick={() => this.changeRange(stat_type)}
                >
                  {stat_name}
                </Pill>
              );
            })}
          </div>
          <div style={{ flex: "initial", marginLeft: "auto" }}>
            {Boolean(userStatsUrl) && (
              <a
                href={userStatsUrl}
                className={`pill secondary ${user ? "active" : ""}`}
              >
                <FontAwesomeIcon icon={faUser} />{" "}
                {user?.name ?? currentUser?.name}
              </a>
            )}
            <a
              href={globalStatsUrl}
              className={`pill secondary ${!user ? "active" : ""}`}
            >
              <FontAwesomeIcon icon={faGlobe} /> Global
            </a>
          </div>
        </div>
        <section id="listening-activity">
          <ErrorBoundary>
            <UserListeningActivity range={range} apiUrl={apiUrl} user={user} />
          </ErrorBoundary>
        </section>
        <section id="top-entity">
          <div className="row">
            <div className="col-md-4">
              <ErrorBoundary>
                <UserTopEntity
                  range={range}
                  entity="artist"
                  apiUrl={apiUrl}
                  user={user}
                  terminology="artist"
                  newAlert={newAlert}
                />
              </ErrorBoundary>
            </div>
            <div className="col-md-4">
              <ErrorBoundary>
                <UserTopEntity
                  range={range}
                  entity="release-group"
                  apiUrl={apiUrl}
                  user={user}
                  terminology="album"
                  newAlert={newAlert}
                />
              </ErrorBoundary>
            </div>
            <div className="col-md-4">
              <ErrorBoundary>
                <UserTopEntity
                  range={range}
                  entity="recording"
                  apiUrl={apiUrl}
                  user={user}
                  terminology="track"
                  newAlert={newAlert}
                />
              </ErrorBoundary>
            </div>
          </div>
        </section>
        {user && (
          <section id="daily-activity">
            <ErrorBoundary>
              <UserDailyActivity range={range} apiUrl={apiUrl} user={user} />
            </ErrorBoundary>
          </section>
        )}
        <section id="artist-origin">
          <ErrorBoundary>
            <UserArtistMap range={range} apiUrl={apiUrl} user={user} />
          </ErrorBoundary>
        </section>
      </div>
    );
  }
}

document.addEventListener("DOMContentLoaded", () => {
  const {
    domContainer,
    reactProps,
    globalAppContext,
    sentryProps,
    optionalAlerts,
  } = getPageProps();
  const { sentry_dsn, sentry_traces_sample_rate } = sentryProps;

  if (sentry_dsn) {
    Sentry.init({
      dsn: sentry_dsn,
      integrations: [new Integrations.BrowserTracing()],
      tracesSampleRate: sentry_traces_sample_rate,
    });
  }
  const { user } = reactProps;
  const UserReportsPageWithAlertNotifications = withAlertNotifications(
    UserReports
  );

  const renderRoot = createRoot(domContainer!);
  renderRoot.render(
    <ErrorBoundary>
      <GlobalAppContext.Provider value={globalAppContext}>
        <NiceModal.Provider>
          <UserReportsPageWithAlertNotifications
            apiUrl={globalAppContext.APIService.APIBaseURI}
            user={user}
            initialAlerts={optionalAlerts}
          />
        </NiceModal.Provider>
      </GlobalAppContext.Provider>
    </ErrorBoundary>
  );
});
