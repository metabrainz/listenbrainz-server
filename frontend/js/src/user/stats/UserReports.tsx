import { createRoot } from "react-dom/client";
import * as React from "react";

import * as Sentry from "@sentry/react";
import { Integrations } from "@sentry/tracing";
import NiceModal from "@ebay/nice-modal-react";
import {
  faGlobe,
  faInfoCircle,
  faUser,
} from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import ErrorBoundary from "../../utils/ErrorBoundary";
import Pill from "../../components/Pill";
import UserListeningActivity from "./components/UserListeningActivity";
import UserTopEntity from "./components/UserTopEntity";
import UserDailyActivity from "./components/UserDailyActivity";
import UserArtistMap from "./components/UserArtistMap";
import { getPageProps } from "../../utils/utils";
import { getAllStatRanges } from "./utils";
import withAlertNotifications from "../../notifications/AlertNotificationsHOC";
import GlobalAppContext from "../../utils/GlobalAppContext";

export type UserReportsProps = {
  user?: ListenBrainzUser;
  apiUrl: string;
};

export type UserReportsState = {
  range: UserStatsAPIRange;
  user?: ListenBrainzUser;
};

export default class UserReports extends React.Component<
  UserReportsProps,
  UserReportsState
> {
  static contextType = GlobalAppContext;
  declare context: React.ContextType<typeof GlobalAppContext>;

  constructor(props: UserReportsProps) {
    super(props);

    this.state = {
      range: "" as UserStatsAPIRange,
      user: props.user,
    };
  }

  componentDidMount() {
    window.addEventListener("popstate", this.syncStateWithURL);

    const range = this.getURLParams();
    window.history.replaceState(
      null,
      "",
      `?range=${range}${window.location.hash}`
    );
    this.syncStateWithURL();
  }

  componentWillUnmount() {
    window.removeEventListener("popstate", this.syncStateWithURL);
  }

  setUser(userName?: string) {
    if (userName) {
      this.setState({ user: { name: userName } });
    } else {
      this.setState({ user: undefined });
    }
  }

  changeRange = (newRange: UserStatsAPIRange): void => {
    this.setURLParams(newRange);
    this.syncStateWithURL();
  };

  syncStateWithURL = async (): Promise<void> => {
    const range = this.getURLParams();
    this.setState({ range });
  };

  getURLParams = (): UserStatsAPIRange => {
    const url = new URL(window.location.href);

    let range: UserStatsAPIRange = "week";
    if (url.searchParams.get("range")) {
      range = url.searchParams.get("range") as UserStatsAPIRange;
    }

    return range;
  };

  setURLParams = (range: UserStatsAPIRange): void => {
    window.history.pushState(null, "", `?range=${range}`);
  };

  render() {
    const { range, user } = this.state;
    const { apiUrl, user: initialUser } = this.props;
    const { currentUser } = this.context;

    const ranges = getAllStatRanges();
    const userOrLoggedInUser: string | undefined =
      user?.name ?? currentUser?.name;

    return (
      <div>
        <div className="tertiary-nav dragscroll">
          <div>
            {Array.from(ranges, ([stat_type, stat_name]) => {
              return (
                <Pill
                  active={range === stat_type}
                  type="secondary"
                  onClick={() => this.changeRange(stat_type)}
                >
                  {stat_name}
                </Pill>
              );
            })}
          </div>
          <div>
            {Boolean(userOrLoggedInUser) && (
              <button
                type="button"
                onClick={() => {
                  this.setUser(initialUser?.name ?? currentUser?.name);
                }}
                className={`pill secondary ${user ? "active" : ""}`}
              >
                <FontAwesomeIcon icon={faUser} />{" "}
                {initialUser?.name ?? currentUser?.name}
              </button>
            )}
            <button
              type="button"
              onClick={() => {
                this.setUser();
              }}
              className={`pill secondary ${!user ? "active" : ""}`}
            >
              <FontAwesomeIcon icon={faGlobe} /> Global
            </button>
          </div>
        </div>
        <small>
          <FontAwesomeIcon icon={faInfoCircle} />
          &nbsp;
          <a
            href="https://listenbrainz.readthedocs.io/en/latest/general/data-update-intervals.html"
            target="_blank"
            rel="noopener noreferrer"
          >
            How often are my stats updated?
          </a>
        </small>
        <section id="listening-activity">
          <ErrorBoundary>
            <UserListeningActivity range={range} apiUrl={apiUrl} user={user} />
          </ErrorBoundary>
        </section>
        <section id="top-entity">
          <div className="row">
            <div className="col-md-4">
              <ErrorBoundary>
                <UserTopEntity
                  range={range}
                  entity="artist"
                  apiUrl={apiUrl}
                  user={user}
                  terminology="artist"
                />
              </ErrorBoundary>
            </div>
            <div className="col-md-4">
              <ErrorBoundary>
                <UserTopEntity
                  range={range}
                  entity="release-group"
                  apiUrl={apiUrl}
                  user={user}
                  terminology="album"
                />
              </ErrorBoundary>
            </div>
            <div className="col-md-4">
              <ErrorBoundary>
                <UserTopEntity
                  range={range}
                  entity="recording"
                  apiUrl={apiUrl}
                  user={user}
                  terminology="track"
                />
              </ErrorBoundary>
            </div>
          </div>
        </section>
        {user && (
          <section id="daily-activity">
            <ErrorBoundary>
              <UserDailyActivity range={range} apiUrl={apiUrl} user={user} />
            </ErrorBoundary>
          </section>
        )}
        <section id="artist-origin">
          <ErrorBoundary>
            <UserArtistMap range={range} apiUrl={apiUrl} user={user} />
          </ErrorBoundary>
        </section>
      </div>
    );
  }
}

document.addEventListener("DOMContentLoaded", async () => {
  const {
    domContainer,
    reactProps,
    globalAppContext,
    sentryProps,
  } = await getPageProps();
  const { sentry_dsn, sentry_traces_sample_rate } = sentryProps;

  if (sentry_dsn) {
    Sentry.init({
      dsn: sentry_dsn,
      integrations: [new Integrations.BrowserTracing()],
      tracesSampleRate: sentry_traces_sample_rate,
    });
  }
  const { user } = reactProps;
  const UserReportsPageWithAlertNotifications = withAlertNotifications(
    UserReports
  );

  const renderRoot = createRoot(domContainer!);
  renderRoot.render(
    <ErrorBoundary>
      <GlobalAppContext.Provider value={globalAppContext}>
        <NiceModal.Provider>
          <UserReportsPageWithAlertNotifications
            apiUrl={globalAppContext.APIService.APIBaseURI}
            user={user}
          />
        </NiceModal.Provider>
      </GlobalAppContext.Provider>
    </ErrorBoundary>
  );
});
