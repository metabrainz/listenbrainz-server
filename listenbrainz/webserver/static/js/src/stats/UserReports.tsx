import * as ReactDOM from "react-dom";
import * as React from "react";

import ErrorBoundary from "../ErrorBoundary";
import Pill from "../components/Pill";
import UserListeningActivity from "./UserListeningActivity";
import UserTopEntity from "./UserTopEntity";
import UserDailyActivity from "./UserDailyActivity";
import UserArtistMap from "./UserArtistMap";

export type UserReportsProps = {
  user: ListenBrainzUser;
  apiUrl: string;
};

export type UserReportsState = {
  range: UserStatsAPIRange;
};

export default class UserReports extends React.Component<
  UserReportsProps,
  UserReportsState
> {
  constructor(props: UserReportsProps) {
    super(props);

    this.state = {
      range: "" as UserStatsAPIRange,
    };
  }

  componentDidMount() {
    window.addEventListener("popstate", this.syncStateWithURL);

    const range = this.getURLParams();
    window.history.replaceState(
      null,
      "",
      `?range=${range}${window.location.hash}`
    );
    this.syncStateWithURL();
  }

  componentWillUnmount() {
    window.removeEventListener("popstate", this.syncStateWithURL);
  }

  changeRange = (newRange: UserStatsAPIRange): void => {
    this.setURLParams(newRange);
    this.syncStateWithURL();
  };

  syncStateWithURL = async (): Promise<void> => {
    const range = this.getURLParams();
    this.setState({ range });
  };

  getURLParams = (): UserStatsAPIRange => {
    const url = new URL(window.location.href);

    let range: UserStatsAPIRange = "week";
    if (url.searchParams.get("range")) {
      range = url.searchParams.get("range") as UserStatsAPIRange;
    }

    return range;
  };

  setURLParams = (range: UserStatsAPIRange): void => {
    window.history.pushState(null, "", `?range=${range}`);
  };

  render() {
    const { range } = this.state;
    const { apiUrl, user } = this.props;

    return (
      <div>
        <div className="row mt-15">
          <div className="col-xs-12">
            <Pill
              active={range === "week"}
              type="secondary"
              onClick={() => this.changeRange("week")}
            >
              Week
            </Pill>
            <Pill
              active={range === "month"}
              type="secondary"
              onClick={() => this.changeRange("month")}
            >
              Month
            </Pill>
            <Pill
              active={range === "year"}
              type="secondary"
              onClick={() => this.changeRange("year")}
            >
              Year
            </Pill>
            <Pill
              active={range === "all_time"}
              type="secondary"
              onClick={() => this.changeRange("all_time")}
            >
              All Time
            </Pill>
          </div>
        </div>
        <section id="listening-activity">
          <ErrorBoundary>
            <UserListeningActivity range={range} apiUrl={apiUrl} user={user} />
          </ErrorBoundary>
        </section>
        <section id="top-entity">
          <div className="row">
            <div className="col-md-4">
              <ErrorBoundary>
                <UserTopEntity
                  range={range}
                  entity="artist"
                  apiUrl={apiUrl}
                  user={user}
                />
              </ErrorBoundary>
            </div>
            <div className="col-md-4">
              <ErrorBoundary>
                <UserTopEntity
                  range={range}
                  entity="release"
                  apiUrl={apiUrl}
                  user={user}
                />
              </ErrorBoundary>
            </div>
            <div className="col-md-4">
              <ErrorBoundary>
                <UserTopEntity
                  range={range}
                  entity="recording"
                  apiUrl={apiUrl}
                  user={user}
                />
              </ErrorBoundary>
            </div>
          </div>
        </section>
        <section id="daily-activity">
          <ErrorBoundary>
            <UserDailyActivity range={range} apiUrl={apiUrl} user={user} />
          </ErrorBoundary>
        </section>
        <section id="artist-origin">
          <ErrorBoundary>
            <UserArtistMap range={range} apiUrl={apiUrl} user={user} />
          </ErrorBoundary>
        </section>
      </div>
    );
  }
}

document.addEventListener("DOMContentLoaded", () => {
  const domContainer = document.querySelector("#react-container");
  const propsElement = document.getElementById("react-props");
  let reactProps;
  try {
    reactProps = JSON.parse(propsElement!.innerHTML);
  } catch (err) {
    // Show error to the user and ask to reload page
  }
  const { user, api_url: apiUrl } = reactProps;
  ReactDOM.render(
    <ErrorBoundary>
      <UserReports apiUrl={apiUrl} user={user} />
    </ErrorBoundary>,
    domContainer
  );
});
