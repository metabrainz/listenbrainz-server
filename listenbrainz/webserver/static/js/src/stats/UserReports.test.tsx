import * as React from "react";
import { shallow } from "enzyme";

import UserReports, { UserReportsProps } from "./UserReports";

const props: UserReportsProps = {
  user: {
    name: "test_user",
  },
  apiUrl: "foobar",
};

describe("UserReports", () => {
  it("renders without crashing", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});

describe("ComponentDidMount", () => {
  it('adds event listener for "popstate" event', () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    const spy = jest.spyOn(window, "addEventListener");
    spy.mockImplementationOnce(() => {});
    instance.syncStateWithURL = jest.fn();
    instance.componentDidMount();

    expect(spy).toHaveBeenCalledWith("popstate", instance.syncStateWithURL);
  });

  it("calls getURLParams once", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    instance.getURLParams = jest.fn();
    instance.syncStateWithURL = jest.fn();
    instance.componentDidMount();

    expect(instance.getURLParams).toHaveBeenCalledTimes(1);
  });

  it("calls replaceState with correct parameters", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    const spy = jest.spyOn(window.history, "replaceState");
    spy.mockImplementationOnce(() => {});
    instance.getURLParams = jest.fn().mockImplementationOnce(() => "week");
    instance.syncStateWithURL = jest.fn();
    instance.componentDidMount();

    expect(spy).toHaveBeenCalledWith(null, "", "?range=week");
  });

  it("calls syncStateWithURL", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    instance.syncStateWithURL = jest.fn();
    instance.componentDidMount();

    expect(instance.syncStateWithURL).toHaveBeenCalledTimes(1);
  });
});

describe("componentWillUnmount", () => {
  it('removes "popstate" event listener', () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    const spy = jest.spyOn(window, "removeEventListener");
    spy.mockImplementationOnce(() => {});
    instance.syncStateWithURL = jest.fn();
    instance.componentWillUnmount();

    expect(spy).toHaveBeenCalledWith("popstate", instance.syncStateWithURL);
  });
});

describe("changeRange", () => {
  it("calls setURLParams with correct parameters", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    instance.setURLParams = jest.fn();
    instance.syncStateWithURL = jest.fn();
    instance.changeRange("year");

    expect(instance.setURLParams).toHaveBeenCalledWith("year");
  });

  it("calls syncStateWithURL once", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    instance.syncStateWithURL = jest.fn();
    instance.changeRange("year");

    expect(instance.syncStateWithURL).toHaveBeenCalledTimes(1);
  });
});

describe("syncStateWithUrl", () => {
  it("calls getURLParams once", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    instance.getURLParams = jest.fn();
    instance.syncStateWithURL();

    expect(instance.getURLParams).toHaveBeenCalledTimes(1);
  });

  it("sets state correcty", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    instance.getURLParams = jest.fn().mockImplementationOnce(() => "month");
    instance.syncStateWithURL();

    expect(wrapper.state("range")).toEqual("month");
  });
});

describe("getURLParams", () => {
  it("gets default parameters if none are provided in the URL", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    delete window.location;
    window.location = {
      href: "https://foobar.org",
    } as Window["location"];
    const range = instance.getURLParams();

    expect(range).toEqual("week");
  });

  it("gets parameters if provided in the URL", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    delete window.location;
    window.location = {
      href: "https://foobar.org?range=year",
    } as Window["location"];
    const range = instance.getURLParams();

    expect(range).toEqual("year");
  });
});

describe("setURLParams", () => {
  it("sets URL parameters", () => {
    const wrapper = shallow<UserReports>(<UserReports {...props} />);
    const instance = wrapper.instance();

    const spy = jest.spyOn(window.history, "pushState");
    spy.mockImplementationOnce(() => {});

    instance.setURLParams("all_time");
    expect(spy).toHaveBeenCalledWith(null, "", "?range=all_time");
  });
});
