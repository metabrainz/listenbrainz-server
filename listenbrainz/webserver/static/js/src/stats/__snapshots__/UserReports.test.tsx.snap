// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserReports renders without crashing 1`] = `
<div>
  <div
    className="row mt-15"
  >
    <div
      className="col-xs-12"
    >
      <Pill
        active={true}
        onClick={[Function]}
        type="secondary"
      >
        Week
      </Pill>
      <Pill
        active={false}
        onClick={[Function]}
        type="secondary"
      >
        Month
      </Pill>
      <Pill
        active={false}
        onClick={[Function]}
        type="secondary"
      >
        Year
      </Pill>
      <Pill
        active={false}
        onClick={[Function]}
        type="secondary"
      >
        All Time
      </Pill>
    </div>
  </div>
  <section
    id="listening-activity"
  >
    <ErrorBoundary>
      <UserListeningActivity
        apiUrl="foobar"
        range="week"
        user={
          Object {
            "name": "test_user",
          }
        }
      />
    </ErrorBoundary>
  </section>
  <section
    id="top-entity"
  >
    <div
      className="row"
    >
      <div
        className="col-md-4"
      >
        <ErrorBoundary>
          <UserTopEntity
            apiUrl="foobar"
            entity="artist"
            range="week"
            user={
              Object {
                "name": "test_user",
              }
            }
          />
        </ErrorBoundary>
      </div>
      <div
        className="col-md-4"
      >
        <ErrorBoundary>
          <UserTopEntity
            apiUrl="foobar"
            entity="release"
            range="week"
            user={
              Object {
                "name": "test_user",
              }
            }
          />
        </ErrorBoundary>
      </div>
      <div
        className="col-md-4"
      >
        <ErrorBoundary>
          <UserTopEntity
            apiUrl="foobar"
            entity="recording"
            range="week"
            user={
              Object {
                "name": "test_user",
              }
            }
          />
        </ErrorBoundary>
      </div>
    </div>
  </section>
  <section
    id="daily-activity"
  >
    <ErrorBoundary>
      <UserDailyActivity
        apiUrl="foobar"
        range="week"
        user={
          Object {
            "name": "test_user",
          }
        }
      />
    </ErrorBoundary>
  </section>
  <section
    id="artist-origin"
  >
    <ErrorBoundary>
      <UserArtistMap
        apiUrl="foobar"
        range="week"
        user={
          Object {
            "name": "test_user",
          }
        }
      />
    </ErrorBoundary>
  </section>
</div>
`;
